Support ngrx - Loading children using external libraries (Redux, ect`)

TMaster opened the pull request:
Hi @rychkog 

I added a support for Redux loading in the tree.

The way it works is the following:

- The `TreeModel` exposes an optional  boolean field called `hasChildren`. if this property is set to true, the `Tree` will **consider the current node as a `branch`** and allow expanding it.
- When the node is expanded an `LoadNextLevel` event is fired (through the `tree.service` to the `tree.component` and from  there **an event is emitted to the user**.
- The user should handle the event and change the state of the tree (load the next level).
- When the redux pipeline is finished the root `TreeModel` is now containing the children and those children will be viewed in the tree as if a `children` property was provided.

2 Things you should know:

1.  If the user proivided a `TreeModel` with `loadChildren` function and `hasChildren` set to true. **The `loadChildren` will take precedence and no `loadNextLevel` event will be fired.**
2. while firing the event **the  `_childrenLoadingState` will transit to `Loading` state.** which means the event will fire once per node (on the second time the loading state will prevent from additional event to be fired).

One small issue which is still open: when the tree is recreated all the nodes are expanded by default. That means if the user collapsed one branch and then expanded another branch for the first time, the tree will be recreated and all the branches (including the collapsed one) will be expanded.

For now I handled it with a bookepper that remember all the collapsed nodes and re collapsed them.
I wish to have a better solution, for example the `TreeModel` will have an `expanded` property that will tell the initial state of the node.
What do you say?

Thanks a lot,
TMaster.

TMaster commented:
Hi @rychkog  could you help me understand what is the problem with codecov?

Is ti want me to write tests for the `LoadNextLevel`, where should I add them.

I Haven't seend any tests for the other emitters.

Can you please help?

Thanks,
TMaster

rychkog commented on src/tree.service.ts:
```diff
@@ -98,4 +107,18 @@ export class TreeService {
+
+  private shouldFireLoadNextLevel(tree: Tree): boolean {
+
+    const result = tree.node.hasChildren &&
```
@TMaster seems like not all the branches are covered in this condition - by adding more tests (until yellow mark turned into green one), you will fix this. Also have a look at [utility functions](https://github.com/valor-software/ng2-tree/blob/master/src/utils/fn.utils.ts)  that you can use here. For example instead of `tree.children === []`, you can use `isEmpty(tree.children)`

TMaster replied:
Hi @rychkog, what do you mean by **until yellow mark turned into green one**, what marks?

Thanks

rychkog replied:
@TMaster Ah, sorry for that :)) I have a codecov plugin that shows me covered lines in the PR :) Try it out - it might be very helpful :)
https://chrome.google.com/webstore/detail/codecov-extension/keefkhehidemnokodkdkejapdgfjmijf

rychkog replied:
![image](https://user-images.githubusercontent.com/1942440/31536490-77e730b6-b007-11e7-8f51-aa0c332656dd.png)

rychkog commented on src/tree.service.ts:
```diff
@@ -98,4 +107,18 @@ export class TreeService {
+      !tree.childrenAreBeingLoaded() &&
+      (!tree.children || tree.children === []);
+
+      if (result) {
```
Probably instead of `result` it might be a bit more convenient to name it smth like `isChildrenLoadingRequested`

rychkog commented on src/tree.ts:
```diff
@@ -337,7 +340,7 @@ export class Tree {
    */
   public isBranch(): boolean {
-    return Array.isArray(this._children);
+    return this.node.hasChildren || Array.isArray(this._children);
```
`this.node.hasChildren === true` for consistency with other changes

rychkog commented on src/tree.ts:
```diff
@@ -408,10 +411,9 @@ export class Tree {
    */
   public switchFoldingType(): void {
-    if (this.isLeaf() || !this.hasChildren()) {
+    if ((this.isLeaf() || !this.hasChildren())) {
```
unnecessary parentheses wrapping - can be removed

rychkog commented on test/tree.spec.ts:
```diff
@@ -1086,4 +1086,32 @@ describe('Tree', () => {
+
+    const model: TreeModel = {
+      value: 'root',
+      hasChildren: true,
```
@TMaster hasChildren property means that we tree has children, but under the hood, it means: "Notify me when I can load more children", right? Maybe we can name it a bit more precisely? What do you think?

TMaster replied:
Exactly, It suppose to mean there is mean `There are more children out there in the back end, just let me know and I will load them`. More suitable name perhaps may be something like:
- deferredChildrenExists. 
- additionalChildrenExists.
- additionalLoadingAvailable.

@rychkog what do you say?

rychkog replied:
Something like (taking advanatge over LoadNextLevelEvent name)
`emitLoadNextLevelEvent: true`
Smth like that. Or maybe `notifyAboutNextLevelChildrenLoading: true`. Just to make it self explanatory - final decision is up to you

rychkog commented:
@TMaster According to your questions regarding tests: never mind for now - I will fix that later on my own - seems like some subtle condition was not covered.

Also, it would be good to squash all the commits into one - this way we'll have a nicer commit history :)

TMaster commented:
Thanks a lot @rychkog  I will address your feedbacks ASAP.

TMaster commented:
Hi @rychkog 
I refactored what you asked for.

Still the codcov saying i should cover the EventEmitter in the tree.component, however I'm not sure where to write the test. I haven't seen tests for the other EventEmitters.

What do you say?

Thanks,
TMaster

rychkog commented:
@TMaster merged! But we should have this feature described in the docs so that other users know it exists.

TMaster commented:
Sure thing, should I go a head and change the Readme file?

And thanks for the merge. when do you plan to release a new version?

rychkog commented:
@TMaster Hi!

> Sure thing, should I go a head and change the Readme file?

Yes, you just need to extend the Readme with the new feature description. Then I will have a look at PR and we are done there - release in one "doc step" from being done :)

> And thanks for the merge
Thanks for your contributions to this project!

TMaster commented:
Hi, I added the missing documentation to the branch. Can you re merge it, or should I open a new PR?
By the way I also added documentation about the startRenaming method in the controller
